feat(search): add support for premium search experiment by feature flag

dskuza opened the pull request:
## Summary

Adds support for a premium search experiment by feature flag. This experiment updates the scopes to online-only by title, tag, or content.

This also fixes a Sync test that was failing after updating the max download count for Saves and Archive.

If the time comes to remove this experiment, I'd suggest removing the three added scopes and resolving any build errors.

## Implementation Details

- Update feature flags to check via Unleash _or_ Braze
- Add three premium experiment scopes to existing `SearchScope`
- Dynamically update titles based on feature flag when search is first opened
- Piggyback off of existing online-only logic to add online-only support, ensuring it only occurs if the scope is a premium experiment scope
- Add `PremiumExperimentSearch` to handle updated query filter
  - Added as a separate file just to keep the logic for experiment search separate from default online search
 
I initially had an idea to turn `SearchViewModel` into a protocol and have two conforming types (default, and experiment), but the refactor for that would've been too great, so I deferred to a simpler solution that utilized the existing view model.

## Test Steps

- [ ] Add yourself to the premium search experiment via Braze feature flag
- [ ] Launch the app
- [ ] When opening Search, you should see "Title", "Tag", and "Content"
- [ ] When searching by "Title", you should see only results that contain that word in the title
- [ ] When searching by Tag, you should see only the items tagged by your search term (single tag only)
- [ ] When searching by Content, you should see the equivalent of performing "All" outside of the experiment
---
- [ ] Remove yourself to the premium search experiment via Braze feature flag
- [ ] Launch the app
- [ ] When opening Search, you should see "Saves", "Archive", and "All"
- [ ] Regression test on the prior search experience

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

Commit:
refactor(search): rename SearchViewModel to DefaultSearchViewModel

Commit:
feat(feature-flags): add support for checking feature flag assignment via Braze

Commit:
feat(search): set up premium search experiment scopes

Commit:
feat(search): setup basis for premium experiment search

Commit:
feat(search): cache premium experiment search by scope and term

Commit:
feat(graphql): regenerate files for updated schema

Commit:
feat(search): add premium experiment support to search service

Commit:
fix(tests): fix build errors for premium search experiment

Commit:
tests(search): add premium search experiment tests for search service

Commit:
tests(search): add premium experiment tests for search view model

Commit:
refactor(search): update premium search experiment scope names

Commit:
fix(tests): update large list total count

The fixtures were required to update their total count after updating the max allowed download for Saves and Archive

bassrock commented on PocketKit/Sources/PocketGraphTestMocks/PageInfo+Mock.graphql.swift:
```diff
@@ -10,10 +10,10 @@ public class PageInfo: MockObject {
   public typealias MockValueCollectionType = Array<Mock<PageInfo>>
 
   public struct MockFields {
-    @Field<String?>("endCursor") public var endCursor
```
You need to keep these as optionals for our tests to work

dskuza replied:
@bassrock have we been manually updating these? This is generated.

bassrock replied:
Yea we have.

bassrock replied:
When i raised this with Apollo, i was pushed back on by them cause we are using mocks in an unsupported way.

Gio2018 replied:
what's the final take on this? Tests seem to be passing though

bassrock commented on PocketKit/Sources/PocketKit/FeatureFlagService.swift:
```diff
@@ -56,12 +58,20 @@ class FeatureFlagService: NSObject, FeatureFlagServiceProtocol {
     func isAssigned(flag: CurrentFeatureFlags, variant: String?) -> Bool {
-        guard let cachedFlag = featureFlags[flag] else {
-            return false
+        func isAssignedViaUnleash(flag: CurrentFeatureFlags, variant: String?) -> Bool {
```
Woah, we can do functions in functions in swift?

dskuza replied:
Yep!

bassrock commented on PocketKit/Sources/PocketKit/Notifications/PocketBraze.swift:
```diff
@@ -56,19 +58,31 @@ class PocketBraze: NSObject {
             // Could expand to include "development"
             let deployment = isTestFlight ? "testflight" : "app_store"
-            self?.braze.user.setCustomAttribute(key: "ios_deployment", value: deployment)
+            braze.user.setCustomAttribute(key: "ios_deployment", value: deployment)
```
btw @nzeltzer i dont think this works, ive noticed you and I are set to appstore in braze :(

bassrock reviewed (commented):
Gonna defer refer main review to @Gio2018 since theres a lot here.

1 Question i have though is are you tracking the impression that the user felt the feature flag? I know Braze has a method for that and it will help us launch an in app survey after they use it x number of times.

dskuza commented:
> 1 Question i have though is are you tracking the impression that the user felt the feature flag?

@bassrock nope, i'll go ahead and add that

Commit:
feat(feature-flags): track feature flag impressions via Braze as necessary

Gio2018 commented on PocketKit/Sources/PocketKit/MyList/SearchItemsList/PremiumExperimentSearch.swift:
```diff
@@ -0,0 +1,80 @@
+import Combine
+import SharedPocketKit
+
+class PremiumExperimentSearch {
```
are we planning to keep this class if the experiment goes well? In this case I'd remove the "experiment" word from the name and call it something else

dskuza replied:
I'm not sure… I think _for now_ it's best to keep "Experiment" added since _currently_ it's an experiment, and should be renamed later if it's anything otherwise.

Gio2018 commented on PocketKit/Sources/PocketKit/MyList/SearchItemsList/PremiumExperimentSearch.swift:
```diff
@@ -0,0 +1,80 @@
+            }
+            return
+        }
+        clear()
```
can you explain/add some comments on why we need to remove and re-add this subscription every time?

dskuza replied:
To be honest, this is a carryover from the `OnlineSearch` class. I'm not sure why it was originally added. 🙃

Gio2018 commented on PocketKit/Sources/PocketKit/MyList/SearchItemsList/SearchViewModel.swift:
```diff
@@ -354,6 +375,33 @@ class SearchViewModel: ObservableObject {
     }
 
+    /// Submit premium experiment search and update `searchState` with proper view
+    /// - Parameter term: the term the user enters in search bar
```
Looks like this part of the documentation could be a leftover?

Gio2018 commented on PocketKit/Tests/SyncTests/Fixtures/large-list-1.json:
```diff
@@ -5,7 +5,7 @@
       "savedItems": {
         "__typename": "[type-name-here]",
-        "totalCount": 1001,
+        "totalCount": 10001,
```
TY!

Gio2018 reviewed (approved):
Works as expected @dskuza ! Left some comments, for you, but other than that looks good!

Commit:
refactor(search): remove experiment naming from types aside from feature flag

Addresses a pull request comment, and makes it a much smaller change if the experiment is replaced with a different check, if the experiment is kept as a feature and removed from being behind a feature flag
